pkg/plugin/connector/builtin/destination.go

```diff
@@ -0,0 +1,129 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builtin
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/conduitio/conduit-connector-protocol/pconnector"
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/plugin/connector"
+	"github.com/rs/zerolog"
+)
+
+// destinationPluginAdapter implements the destination plugin interface used
+// internally in Conduit and relays the calls to a destination plugin defined in
+// conduit-connector-protocol (pconnector). This adapter needs to make sure it
+// behaves in the same way as the standalone plugin adapter, which communicates
+// with the plugin through gRPC, so that the caller can use both of them
+// interchangeably.
+// All methods of destinationPluginAdapter use runSandbox to catch panics and
+// convert them into errors. The adapter also logs the calls and clones the
+// request and response objects to avoid any side effects.
+type destinationPluginAdapter struct {
+	impl pconnector.DestinationPlugin
+	// logger is used as the internal logger of destinationPluginAdapter.
+	logger log.CtxLogger
+	// ctxLogger is attached to the context of each call to the plugin.
+	ctxLogger zerolog.Logger
+}
+
+var _ connector.DestinationPlugin = (*destinationPluginAdapter)(nil)
+
+func newDestinationPluginAdapter(impl pconnector.DestinationPlugin, logger log.CtxLogger) *destinationPluginAdapter {
+	return &destinationPluginAdapter{
+		impl:      impl,
+		logger:    logger.WithComponent("builtin.destinationPluginAdapter"),
+		ctxLogger: logger.WithComponent("plugin").ZerologWithComponent()}
+}
+
+func (s *destinationPluginAdapter) withLogger(ctx context.Context) context.Context {
+	return s.ctxLogger.WithContext(ctx)
+}
+
+func (s *destinationPluginAdapter) Configure(ctx context.Context, in pconnector.DestinationConfigureRequest) (pconnector.DestinationConfigureResponse, error) {
+	s.logger.Debug(ctx).Any("request", in).Msg("calling Configure")
+	out, err := runSandbox(s.impl.Configure, s.withLogger(ctx), in.Clone(), s.logger)
+	return out.Clone(), err
+}
+
+func (s *destinationPluginAdapter) Open(ctx context.Context, in pconnector.DestinationOpenRequest) (pconnector.DestinationOpenResponse, error) {
+	s.logger.Debug(ctx).Any("request", in).Msg("calling Open")
+	out, err := runSandbox(s.impl.Open, s.withLogger(ctx), in.Clone(), s.logger)
+	return out.Clone(), err
+}
+
+func (s *destinationPluginAdapter) Run(ctx context.Context, stream pconnector.DestinationRunStream) error {
```

Comment:
Nitpick: `d` would be a better receiver name.

Reply:
copy pasta 🍝